[RFC] Java Camera2 View for Android

floe opened the pull request:
Based on #10050, here's a first alpha implementation of the Camera2 adapter class which should allow zero-copy access to preview image data. Comments very welcome.

Known issues at the moment:

- ~~tutorial2-mixedprocessing works with this class as a drop-in replacement for `JavaCameraView`, but quits after ~ 20-30 seconds (no crash, just exit due to internal camera queue error, perhaps some race condition?)~~ - fixed, was a memory leak
- lots of code overlap with `JavaCameraView` and `Camera2Renderer` (e.g. `JavaCameraFrame` internal class), needs cleanup/refactoring
- `cacPreviewSize(...)` currently buggy, selects smaller preview size than feasible

```
force_builders=Android pack
```

alalek commented on modules/java/generator/src/java/android+JavaCamera2View.java:
```diff
@@ -0,0 +1,366 @@
+                    if (image == null) return;
+                    ByteBuffer plane = image.getPlanes()[0].getBuffer();
+                    Mat yuv = new Mat( h * 3/2, w, CvType.CV_8UC1, plane );
+                    JavaCamera2Frame tempFrame = new JavaCamera2Frame(yuv,w,h);
```
[This](https://developer.android.com/reference/android/media/Image.html#getFormat()) says that all three planes should be used for `YUV_420_888`.

BTW, Please add additional sanity checks (assertions) - for actual `Image` format, `ByteBuffer` size, etc.

floe replied:
Hmm... OTOH, you're correct (there are 3 planes), but obviously, the planes are contiguous in memory and also mapped in one single chunk, otherwise, the `cvtColor` call would cause a segfault. 

To implement this in a way that references all 3 planes separately would AFAICT require a new implementation of `cvtColor` that can handle 3 separate `Mat` objects. Otherwise, you'd need to copy the planes together into a single `Mat` and lose the advantage of the zero-copy native buffer.

mshabunin replied:
We have `hal::` conversion functions accepting two separate pointers for Y and UV planes ([example](https://github.com/opencv/opencv/blob/ba940dec38983cb35de6342d5e770eb0f054549e/modules/imgproc/include/opencv2/imgproc/hal/hal.hpp#L195-L199), so you can try to use them. You can add more overloaded functions if needed. But it is hard to say how it will work taking in account that each plane has its own `PixelStride` parameter which is not supported by current implementations.

floe replied:
According to https://developer.android.com/reference/android/graphics/ImageFormat.html#YUV_420_888 the Y plane will always have a stride of 1, and the U/V planes will have equal stride. I may simply assert that U/V stride is also 1 for the moment (seems to be true for all implementations I've seen so far).

mshabunin replied:
Are you sure U and V have PixelStride=1? Probably it is 2. If your current code has worked with YUV_BGRA_NV12 conversion by providing only _Y_ pointer, then planes should've been located in continuous memory block like this:
```
YYYY
YYYY
UVUV
```
Then: 
```
Uptr = Yptr + row_stride * rows
Upixel_stride = 2
Vptr = Uptr + 1
Vpixel_stride = 2
```

mshabunin replied:
Corrections to my previous comment: the conversion you used is YUV2RGB2_NV21 and in memory layout the V channel stands before U.

floe replied:
But I'm using COLOR_YUV2RGB_YV12 (see below), and that is for 3 sequential planes, isn't it?

mshabunin replied:
As I can see from the code:
```.cpp
private int mPreviewFormat = ImageFormat.NV21;
```
This field is not modified anywhere and is used to determine conversion mode.

floe replied:
My mistake, you're right. So it's one separate Y plane and one interleaved U/V plane.

floe replied:
But in that case, `cvtTwoPlaneYUVtoBGR` should probably work?

alalek replied:
Probably it should, but we should check these assumptions (via `CV_Assert`) before calling any processing code.

mshabunin replied:
@floe , the problem with `cvtTwoPlaneYUVtoBGR` is that it does not get wrapped automatically, because it accepts raw C++ pointers. So you will need to write JNI wrapper by hand or wait_for/propose extended `cvtColor(InputArray ...)` interface.

floe replied:
Is there any way to check whether the planes are contiguous in memory before passing them to `Mat()`? Then the current approach would already be sufficient...

alalek commented on modules/java/generator/src/java/android+JavaCamera2View.java:
```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    boolean cacPreviewSize(final int width, final int height) {
```
There is typo (in original code too): `calcPreviewSize`

floe replied:
Aaah. This is supposed to mean "calc". :man_facepalming: Will fix that.

alalek commented:
Please update files filtering for old Android versions [here](https://github.com/opencv/opencv/blob/3.3.1/modules/java/CMakeLists.txt#L240)

floe commented:
I've updated this PR:

- fixed memory leak
- fixed some typos
- fixed incorrect format handling
- added lots of asserts

The only thing that remains open IMHO is the handling of the RGB conversion here: https://github.com/opencv/opencv/pull/10081/files#diff-980f99834f5223fab7fde83308fc9930R353 

AFAICT the cleanest way would be to store Y and U/V plane as two `Mat` objects in `JavaCamera2Frame` and add a wrapper to `ImgProc.cvtColor` that calls through to `cvtTwoPlaneYUVtoBGR`. Opinions?

floe commented:
Side question: where are the Java wrappers for `cvtColor` defined?

alalek commented:
Almost all wrapper for Java/Python bindings are generated automatically from public headers of modules (functions with `CV_EXPORTS_W`).

BTW, `cv::Mat` doesn't support multiple planes (you need 2 `cv::Mat` to store Y and UV planes).

floe commented:
Another side question: I added
```
force_builders=Android pack
```
to the initial PR comment, but that still doesn't seem to trigger the Android pack build?

floe commented:
*bump* sorry for the noise, but I do need the Android pack to work for testing...

floe commented:
Nevermind, seems to work now :+1:

floe commented:
I think this should now be ready for merging. I still can't get the "Android pack" build to trigger reliably, though. Any comments welcome /cc @alalek @mshabunin @vpisarev

floe commented:
Update: fixed one more bug in the C wrapper; seems to work quite smoothly now. I've tested on an Nexus 5 with Android 7.1.2, and got a reduction in CPU load of about 25% with tutorial-2-mixedprocessing.

floe commented:
One more question: I've called the C wrapper function `cvtColorTwoPlane`. To avoid namespace pollution, would it perhaps be better to just overload `cvtColor`?

alalek commented on modules/imgproc/src/color.cpp:
```diff
@@ -11050,6 +11050,41 @@ inline bool isFullRange(int code)
+        case COLOR_YUV2BGRA_NV21: case COLOR_YUV2RGBA_NV21: case COLOR_YUV2BGRA_NV12: case COLOR_YUV2RGBA_NV12:
+            break;
+        default:
+            return;
```
> return;

This is an error case. You should use `CV_Error()` here.

floe replied:
OK, fixed.

alalek commented on modules/imgproc/include/opencv2/imgproc.hpp:
```diff
@@ -3673,6 +3673,8 @@ channels is derived automatically from src and code.
  */
 CV_EXPORTS_W void cvtColor( InputArray src, OutputArray dst, int code, int dstCn = 0 );
 
+CV_EXPORTS_W void cvtColorTwoPlane( InputArray src1, InputArray src2, OutputArray dst, int code, int dcn = 0 );
```
> dcn

Why? RGB/RGBA codes are different, so we can properly detect "dcn" value like `cvtColor` does.

floe replied:
Ah, right. How about the name? Should I just overload `cvtColor`? After thinking about it a bit more, that actually seems cleaner to me...

floe replied:
OK, fixed. Method name still as-is.

alalek commented on modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java:
```diff
@@ -0,0 +1,393 @@
+                Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4);  // COLOR_YUV2RGBA_YV12 produces inverted colors
+            else if (mPreviewFormat == ImageFormat.YUV_420_888) {
+                assert(mUVFrameData != null);
+                Imgproc.cvtColorTwoPlane(mYuvFrameData,mUVFrameData,mRgba,Imgproc.COLOR_YUV2RGB_NV21, 4);
```
```.diff
-COLOR_YUV2RGB_NV21
+COLOR_YUV2RGBA_NV21
```

floe replied:
OK, fixed.

alalek commented on modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java:
```diff
@@ -0,0 +1,393 @@
+                    Mat y_mat = new Mat( h, w, CvType.CV_8UC1, y_plane );
+                    Mat uv_mat = new Mat( h/2, w, CvType.CV_8UC1, uv_plane );
+                    JavaCamera2Frame tempFrame = new JavaCamera2Frame(y_mat,uv_mat,w,h);
+                    deliverAndDrawFrame(tempFrame);
```
Original `JavaCameraView` interface is asynchronous.
Image processing must be performed in a separate thread, see `CameraWorker`. This callback should just update the latest frame for next processing (non-processed frames are dropped).

floe replied:
Yes, but the `onImageAvailable` listener will already be called within the `mBackgroundHandler` thread. Isn't this equivalent?

alalek replied:
Thread is "background" but it may block camera frames fetching and cause camera timeout/disconnection.
I have no experience with new Camera2 API, could you add some documentation link about this (callback processing/requirements)?

floe replied:
See here: https://developer.android.com/reference/android/media/ImageReader.html#acquireLatestImage()
Since I'm calling `acquireLatestImage`, the queue will be emptied and only the most recent image will be passed on for processing. The documentation also mentions that "This function is recommended to use over acquireNextImage() for most use-cases, as it's more suited for real-time processing." 
To be on the safe side, I'll test what happens if the processing takes a really long time (let's say 500 ms).

alalek commented on modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java:
```diff
@@ -0,0 +1,393 @@
+                    ByteBuffer y_plane = planes[0].getBuffer();
+                    ByteBuffer uv_plane = planes[1].getBuffer();
+                    Mat y_mat = new Mat( h, w, CvType.CV_8UC1, y_plane );
+                    Mat uv_mat = new Mat( h/2, w, CvType.CV_8UC1, uv_plane );
```
```.diff
-Mat( h/2, w, CvType.CV_8UC1, uv_plane );
+Mat(h/2, w/2, CvType.CV_8UC2, uv_plane);
```

floe replied:
OK, fixed.

floe commented on modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java:
```diff
@@ -0,0 +1,393 @@
+                    "Interrupted while createCameraPreviewSession", e);
+        }
+        finally {
+            //mCameraOpenCloseLock.release();
```
It was already part of the original `JavaCameraView`, so I've kept it. However, it's quite possible that this is a kludge from the Camera1 API that's no longer needed for Camera2. I'll see if I can just get rid of it (I'd assume yes ;-)

alalek replied:
Agreed.
Looks like locks are no longer needed (frames are processed in "sync" with callback with Camera2 API).

floe commented:
Removed the old semaphore and ran another test; seems to work just as well. Ready for merging IMHO.

floe commented:
Hm. The Linux OpenCL build failed, looks like a transient build error?

alalek commented:
Build failure is not related to your patch.

I run some tests with `JavaCamera2View` interface via OpenCV "tutorial2" sample and observing problem with application pause/resume events processing (pressing Android 'Home' button). Does it work for you (pause/restore handling)?

floe commented:
I'll have to test that again. What would be the expected behaviour with the old `JavaCameraView`?

alalek commented:
> expected behaviour

Application activity pause/resume handling should work with app crash/closing

alalek commented:
@floe I pushed several updates to this patch. Please take a look.

floe commented:
Thanks for your help!

PhilLab commented:
@floe Awesome, can't wait to try it. Do you know if the current implementation respects the camera orientation? https://github.com/opencv/opencv/issues/9269 and https://github.com/opencv/opencv/issues/8866

floe commented:
@PhilLab sorry for the late reply, have in fact not tried this yet. But since the code is based on `CameraBridgeViewBase` too, I suspect it will have the same issues.

ValkA commented:
I have an Android device that only works with `ImageReader` of `ImageFormat.YV12`.
The output is 3 planes as following:

```
Mat y_mat = new Mat(h, w, CvType.CV_8UC1, image.getPlanes()[0].getBuffer());
Mat u_mat = new Mat(h/2, w/2, CvType.CV_8UC1, image.getPlanes()[1].getBuffer());
Mat v_mat = new Mat(h/2, w/2, CvType.CV_8UC1, image.getPlanes()[2].getBuffer());
```

Do you have an idea how to convert all these into a full sized RGBA without copying/merging matrices?

ValkA commented:
ok inspired by JavaCameraView, this one listener handles both image formats :)
```
ImageReader.OnImageAvailableListener listener = new ImageReader.OnImageAvailableListener {
        int w,h;
        public OnImageAvailableListener(int w, int h){
            this.w = w;
            this.h = h;
        }

        final Mat rgba = new Mat(h,w,CvType.CV_8UC4);

        @Override
        public void onImageAvailable(ImageReader imageReader) {
            Image image = imageReader.acquireLatestImage();
            if(image == null) return;

            Mat yuv420 = new Mat(h+h/2, w, CvType.CV_8UC1, image.getPlanes()[0].getBuffer());
            switch (imageReader.getImageFormat()){
                case ImageFormat.YV12:
                    Imgproc.cvtColor(yuv420, rgba, Imgproc.COLOR_YUV2RGBA_YV12, 4);
                    break;
                case ImageFormat.YUV_420_888:
                    Imgproc.cvtColor(yuv420, rgba, Imgproc.COLOR_YUV2BGR_NV12, 4);
                    break;
            }

            //Do your thing here...

            Utils.matToBitmap(rgba, bitmap);
            image.close();
            cameraView.setImageBitmap(bitmap);
        }
    }
```

ValkA commented:
To fix color conversion bugs (which I had in various devices/emulators) I would propose to do the following 

in JavaCamera2View I would change JavaCamera2Frame to:

```
    private class JavaCamera2Frame implements CvCameraViewFrame {
        @Override
        public Mat gray() {
            return new Mat(mImage.getHeight(), mImage.getWidth(), CvType.CV_8UC1, mImage.getPlanes()[0].getBuffer());
        }

        @Override
        public Mat rgba() {
            Mat yuv420 = new Mat((3*mImage.getHeight())/2, mImage.getWidth(), CvType.CV_8UC1, mImage.getPlanes()[0].getBuffer());
            switch (mImage.getFormat()){
                case ImageFormat.YV12:
                    Imgproc.cvtColor(yuv420, mRgba, Imgproc.COLOR_YUV2RGBA_YV12, 4);
                    break;
                case ImageFormat.NV21:
                    Imgproc.cvtColor(yuv420, mRgba, Imgproc.COLOR_YUV2RGBA_NV21, 4);
                    break;
                case ImageFormat.YUV_420_888:
                    Imgproc.cvtColor(yuv420, mRgba, Imgproc.COLOR_YUV2RGBA_I420, 4);
                    break;
                default:
                    throw new IllegalArgumentException("Preview Format can be NV21 or YV12");
            }

            return mRgba;
        }

        public JavaCamera2Frame(Image image) {
            super();
            mImage = image;
            mRgba = new Mat();
        }

        public void release() {
            mRgba.release();
        }

        private Mat mRgba;
        private Image mImage;
    };
```

and OnImageAvailableListener to:
```
mImageReader.setOnImageAvailableListener(new ImageReader.OnImageAvailableListener() {
                @Override
                public void onImageAvailable(ImageReader reader) {
                    long t0 = System.currentTimeMillis();

                    Image image = reader.acquireLatestImage();
                    if (image == null)
                        return;

                    JavaCamera2Frame tempFrame = new JavaCamera2Frame(image);
                    deliverAndDrawFrame(tempFrame);
                    tempFrame.release();
                    image.close();

                    long t1 = System.currentTimeMillis();
                    Log.d(LOGTAG, "" + 1000/(t1-t0) + "fps");
                }
            }, mBackgroundHandler);
```

ansman commented:
@ValkA Your code only uses the Y plane which would lead to a gray scale picture.

ValkA commented:
It takes also UV in the extra h/2 rows

ansman commented:
Ah, I see. It does seem to be continuous but that is probably a dangerous assumption that can lead to strange issues down the line.

ValkA commented:
Yeah, also when lines are padded it has issues

ansman commented:
@ValkA, @floe, @alalek I've done some experiments and it seems that if you use a different size for your image reader than your surface (or perhaps it's just different from the standard 1920x1080) you don't actually get NV21 because of padding between the rows.

Furthermore it seems that V always comes before U in memory so just taking the second plane as an `uv` buffer seems incorrect (if anything it should take the third plane and treat it as `vu`)

alalek commented:
@ansman Could you please check this patch on your device: #14505 ?

ansman commented:
@alalek I added a review for the PR
